Sort : keep nullable property for sorted segments

j-tr opened the pull request:
fixes #2158

When executing `./hyriseBenchmarkTPCH -e Unencoded` twice, the second run failed while loading the cached tables.
This is not due to the binary reader/writer but due to the sorting of the order and lineitem table.
The sort operator kept the nullability property for the columns but always set all segments to be nullable. This missmatch between column and semgent nullability caused the binary reader/writer to mess up.

j-tr commented:
out of curiosity, i added a debug assert to the append_chunk method in table.cpp that checks if nullable segments are added if the corresponding column is set non-nullable and found out that the AggregateSort suffers from the same problem as the Sort operator.
so exporting and importing tables that have been a result of an aggregate could potentially fail as well. don't know if thats critical

mrks commented on src/test/lib/operators/sort_test.cpp:
```diff
@@ -70,6 +70,40 @@ TEST_P(SortTest, Sort) {
   }
 }
 
+bool column_nullablility_equals_segment_nullability(const std::shared_ptr<const Table>& table) {
```
If it's only used for one test, move it into a lambda?

mrks reviewed (commented):
Thanks for tracking this down!

> This missmatch between column and semgent nullability caused the binary reader/writer to mess up.

Any chance we can add an assertion there that is more speaking than the current error?

> I added a debug assert to the append_chunk method in table.cpp that checks if nullable segments are added if the corresponding column is set non-nullable

Sounds like an assertion that we should add. Either here or a new issue. I don't think all operators use `append_chunk`, maybe make it part of `AbstractOperator::execute`? The other way (column nullable but segment is not) should still be allowed.

j-tr commented:
> The other way (column nullable but segment is not) should still be allowed.

in this case, we need to adapt the BinaryParser. Currently, the parser assumes that all segments have the same nullability as the column.

mrks commented:
Ok, let's not touch this for now and require the segment and column nullability to match. I'd still add a comment explaining that it would be possible to loosen this in the future if we think that it makes sense.

mrks commented:
Are you planning to look into the assertion, too? Otherwise, please make an issue out of it so that we can merge this PR without losing track of it.

j-tr commented:
i implemented the assertion but it failed surprisingly often. now investigating how to fix this. moved it to a new issue since this might be a little bit more complicated and think that this PR can be merged
